Update API to send storage state

vk-en opened the pull request:
During the [discussion](https://github.com/lf-edge/eve/pull/2484#discussion_r813207908), it turned out that information about the state of the storage system should be sent in an information message. This PR carries information about the state of the storage system into an informational message.

eriknordmark commented on api/proto/metrics/metrics.proto:
```diff
@@ -173,6 +173,7 @@ message appCpuMetric {
   uint64 total_ns = 7;                 // like 'total' but in nanoseconds. Overflow may happen after ~40 years without EVE restart
 }
 
+// StorageStatus - DEPRECATED. DO NOT USE THIS. Moved to info
```
Has any code been written in zedcloud or eden? It was added just a week ago, right? So we might be able to just delete the added messages.

giggsoff replied:
We did not include it into Eden and I cannot see any entries on zedcloud side.

vk-en replied:
As for the delete,  there are problems with yetus, and here, in order to avoid them, is it possible to leave it as it is for now? Or is there another way to solve these problems?
> Has any code been written in zedcloud or eden? It was added just a week ago, right? So we might be able to just delete the added messages.

eriknordmark replied:
We can ignore those yetus complaints if we are convinced that nothing has started using the new API. (We judiciously ignore yetus complaints - after careful review etc).

eriknordmark commented on api/proto/metrics/metrics.proto:
```diff
@@ -236,7 +239,7 @@ message deviceMetric {
   // of network connectivity.
   uint64 dormant_time_in_seconds = 22;
-  repeated StorageMetric storage_metrics = 23;
+  repeated StorageMetric storage_metrics = 23; // Deprecated. Do not use this. Moved to info
```
This one can also be deleted and replaced with a // Depreated tag 23
comment.

eriknordmark reviewed (commented):
Some nits.

But shouldn't there also be some storage metrics with counters (per pool and per disk) with read, write, resilver, scrub etc counters? If so it might make sense to leave a relatively empty StorageMetric message with tag 1 and 2, and comments to mark tag 3 and 4 as deprecated.
If so

vk-en commented:
@eriknordmark There are such metrics, but do we really need this information and how do we want to use it?

> Some nits.
> 
> But shouldn't there also be some storage metrics with counters (per pool and per disk) with read, write, resilver, scrub etc counters? If so it might make sense to leave a relatively empty StorageMetric message with tag 1 and 2, and comments to mark tag 3 and 4 as deprecated. If so

eriknordmark commented:
> In addition, I do not think that this is necessary and super important information, so we can add such metrics later in a rarer message.

Agreed it can be added later. But that means we want to keep the StorageMetrics message definition around so we can later extend it.

eriknordmark commented on api/proto/info/info.proto:
```diff
@@ -405,6 +405,27 @@ enum ZDeviceState {
+}
+
+message StorageDiskState {
+  string disk_name = 1;
```
Woud make sense to add a comment indicating what this is e.g., a /dev/sdc name, a disk3 name.
Also, shouldn't we added serial string here as well?

eriknordmark replied:
See comment below about using a evecommon DiskDescription here.

eriknordmark commented on api/proto/metrics/metrics.proto:
```diff
@@ -236,7 +221,7 @@ message deviceMetric {
   // of network connectivity.
   uint64 dormant_time_in_seconds = 22;
-  repeated StorageMetric storage_metrics = 23;
+  // Depreated tag 23
```
I think it makes sense to keep this even if we currently do not fill in the StorageMetrics.

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -352,12 +352,18 @@ func PublishDeviceInfoToZedCloud(ctx *zedagentContext) {
+				rStorageInfo.ZpoolState = new(info.StorageState)
 				curentRaid := info.StorageRaidType_STORAGE_RAID_TYPE_NORAID
 				zpoolName := zpool.Properties[libzfs.PoolPropName].Value
+				storageState := zfs.GetZfsDeviceStatusFromStr(zpool.Properties[libzfs.PoolPropHealth].Value)
```
Can there be odd cases when zpool.Properties[libzfs.PoolPropHealth] does not exist i.e., would it make sense to guard against that being nil?

eriknordmark replied:
But conservative programming says one should check for such things to avoid a golang panic due to the key not being in the map.

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -352,12 +352,18 @@ func PublishDeviceInfoToZedCloud(ctx *zedagentContext) {
 				if err != nil {
 					log.Errorf("error with ParseUint for get zpool size in byte: %v", err)
 				}
+				countZvolume, err := zfs.GetZfsCountVolume(zpool.Properties[libzfs.PoolPropName].Value)
```
Ditto about adding a check for nil zpool.Properties[libzfs.PoolPropName]

vk-en replied:
I do not think that if we are on this line of code, the pool name will not be available to us or have nil.
But still. even if something goes wrong, we'll return 0 since we can't find a dataset for the nil value.

eriknordmark replied:
It isn't about the pool name; it is about avoiding a potential golang panic and instead logging/reporting an error.

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -369,15 +375,34 @@ func PublishDeviceInfoToZedCloud(ctx *zedagentContext) {
+						if vdev.Type == libzfs.VDevTypeMirror || vdev.Type == libzfs.VDevTypeRaidz {
+							for _, disk := range vdev.Devices {
+								rDiskStatus, err := zfs.GetZfsDiskAndStatus(disk)
+								if err == nil {
```
Please log.Errorf if you get an error.

vk-en replied:
In this case, we do not need to display errors if we hit, for example, a snapshot or cache.
If the sent data is not a physical disk, this is normal and possible. In this case, if it is not a disk, then we do not add it.

eriknordmark replied:
Can you please add a comment about this - that we ignore errors since they appear when we have e.g., a snapshot

eriknordmark replied:
That comment isn't useful because it just repeats what the code says. The needed part is that vdev.Devices might includes snapshots or caches, and those will result in errors which we need to ignore.

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -369,15 +375,34 @@ func PublishDeviceInfoToZedCloud(ctx *zedagentContext) {
+						}
+						// If there is no RAID or mirror, add a disk if it is a disk
+						rDiskStatus, err := zfs.GetZfsDiskAndStatus(vdev)
+						if err == nil {
```
Please log.Errorf if you get an error.

eriknordmark replied:
Any update? Any reason to not log an error here? If so add a comment which helps the reader understand why errors need to be ignored.

zed-rishabh commented on api/proto/info/info.proto:
```diff
@@ -405,6 +405,27 @@ enum ZDeviceState {
+}
+
+message StorageState {
+  StorageStatus storage_state = 1;      // zfs volume status
```
@petr-zededa Any specific reason we are not keeping pool state on top level of ```StorageInfo``` structure?

petr-zededa replied:
Good point. It seems we can move content of `StorageState` to the top of `StorageInfo` and remove `StorageState` to reduce complexity. WDYT @vk-en ?

zed-rishabh commented on api/proto/info/info.proto:
```diff
@@ -451,6 +472,8 @@ message StorageInfo {
   StorageRaidType curent_raid = 4;  // Current RAID configuration. When no have RAID send ""
```
Please fix typo

vk-en commented:
@zed-rishabh @eriknordmark Corrected and update

vk-en commented:
Thank you all for your comments. I corrected the checks for nil when working with properties. Fixed a bug that zed-rishabh noticed. Removed the division of structures into status and information, placing everything in the information. Added comments in some places that were requested earlier.

giggsoff reviewed (approved):
LGTM

eriknordmark commented on api/proto/info/info.proto:
```diff
@@ -405,6 +405,22 @@ enum ZDeviceState {
+}
+
+message StorageDiskState {
+  string disk_name = 1;                // for example /dev/sda
```
I think we should also have a logical name and a serial here to make sure the zmsg and UI work is prepared to handle the different way to identify the disks the same way for info and for config. https://github.com/lf-edge/eve/pull/2531 has a disk_description and maybe that should be added to evecommon so we can share it between info and config subdirs.

eriknordmark replied:
Note that Petr has updated #2531 so you can import the disk description from evecommon.

eriknordmark commented on pkg/pillar/cmd/zedagent/reportinfo.go:
```diff
@@ -351,33 +351,79 @@ func PublishDeviceInfoToZedCloud(ctx *zedagentContext) {
 				}
 
-				zpool.Close()
+				defer zpool.Close()
```
This is the wrong place of a defer. Normally a defer close is immediately after the open. Since here you do a ZpoolOpenAll then presumably the defer close should be at the top of the loop. Note that there are continue statements above hence this is more than a stylistic comment.

vk-en commented:
I opened a [new PR #2535](https://github.com/lf-edge/eve/pull/2535)  to catch up with the master branch and avoid build conflicts.
